### ISSUE TITLE
More information about Rucio failures would be useful in PostJob

### ISSUE THREAD

**Issue by KatyEllis** [dmwm/CRABServer#7303]
Looking at the PostJob, there is no information from Rucio/FTS about why the output data transfer failed. During my test, I was able to find this information in Kibana, but maybe a User could not do this, and it would be helpful to propagate the reason for failure to the PostJob. 

For example, during my test, the transfer to Tier 2 site failed multiple times due to permission errors:
TRANSFER [1] DESTINATION MAKE_PARENT HTTP 403 : Permission refused

**Comment by dciangot** [comment]
#6041

**Comment by mapellidario** [comment]
I submitted two tasks with rucio stageout enabled:

- `220524_162113:dmapelli_crab_20220524_182110`, rucio rule ID for stageout: `97a981bd7a97477d8fd4f284a1daea8b`
- `220608_085455:dmapelli_crab_20220608_105452`, rucio rule ID for stageout: `67a9e9305cab4300ab1337ef77d234bf`

These tasks use the same configuration, 10 jobs. In both cases, 5 files have been transferred successfully, 5 transfer were stuck.

Together with @KatyEllis , checked the logs of every rucio transfer attempt an noticed some quirks that we may want to address.

- The rucio rule related to the first task (2022-05-24) has been trying to copy the 5 files over and over since when I run `crab resubmit`. https://monit-kibana.cern.ch/kibana/goto/2f4b3d687bc9e25335419f46bfd27705 The logs state that the error is the "source checksum mismatch" that we know about. The question is: is it reasonable that we as crab marked thos files as "failed" because after 24h the postjob gives up trying to transfer the output, but rucio under the hood has been tirelessly trying to transfer it? Can we do better? can we avoid this? should we put a lifetime to stageout the rucio rules? Eventually I asked Katy to delete this rule, she had the permissions to do so
- The rucio rule related to the second task (2022-06-08) experienced something similar. the transfer of 5 output files is marked as stuck after the `crab resubmit`, but rucio is continuously trying to transfer the output of job 10 https://monit-kibana.cern.ch/kibana/goto/5c8fdcbe96c510274bab3301f2637e30 Why only this? why not try to transfer the output of all the jobs that we marked as failed? 
- I managed to delete the rule related to the second task, but i do not have the rights to change its lifetime. Let's see how long it takes for rucio to permanently delete it.

```plaintext
> [17:40] lxplus8s15 ~
> rucio delete-rule 67a9e9305cab4300ab1337ef77d234bf
> [17:40] lxplus8s15 ~
> rucio update-rule --lifetime 0 67a9e9305cab4300ab1337ef77d234bf
2022-06-21 17:41:25,456 ERROR   Access to the requested resource denied.
Details: Account dmapelli can not update this replication rule.
```

Long story short: we may want to improve a bit the tracking of the underlying rucio rules in our stageout script and possibly even manipulate the rules after we create them. I know this is a vague statement, I know this is not trivial and i know this require some experience with rucio, definitely more than I have at the moment. As with every other scientific paper, "more ~~experiments~~ tests on this subject will be needed".

**Comment by belforte** [comment]
information is one thing. Design faults another. We know that `We currently have a problem with Rucio stageout in case of  job resubmissions after a file was produced` (from meeting minutes). And that needs to be addressed of course (even if I do not have a solution in mind !).
But what worries me now is rule ownership.
We should not need to reach out to Rucio admins (Katy) to remove rules.

As to "we call if failed but Rucio keeps trying"... very good point, if we decide to give up on waiting for Rucio, we should remove the current rule as well so that state output files left around in `tmp` will be garbage collected eventually.

Sort of .. too many things for this issue !! Let's move those points to an ad hoc document.

**Comment by dciangot** [comment]
FYI regarding the first point I just started a dedicated discussion on https://github.com/dmwm/CMSRucio/issues/297

**Comment by KatyEllis** [comment]
We found that a user (e.g. Dario) could delete a rule (although I thought that @dciangot mentioned this was not possible?). I also showed Dario how to update a rule, but this is not needed to delete the rule, I only used this for purposes of speeding up the cessation of the rule trying to make the transfer. It makes sense for users not to be able to update, as they could extend the lifetime of the rule beyond that approved by the site admin. So Users can delete rules for failing transfers, but will they?

The other point about Rucio continuing to attempt transfers after CRAB has lost interest is very valid and indeed a different issue to the one reported in this ticket.

**Comment by belforte** [comment]
well.. if editing a rule in not possible, we can always make a new one as long as deletion works, so I'd call that "solved". The crux of the problem seems that CRAB counts time from when transfer started, for each job. Rucio counts time from when rule was created, same for all job, IIUC. Maybe CRAB PostJob should not timeout on transfer anymore, but watch for STUCK replications, the problem is "what do we do when a lock becomes STUCK" ? Do we try to rescue things with possibly a new submission, or wait for DT operators to do something ? Currently PostJob waits 24h and then resubmits, maybe it should wait a week and then mark as failed ?
(users can resubmit, and hit the checksum mismatch problem, let's address that separately).

Do we wait and see how often this happens ?
If user does  not type `crab resubmit`, how often we end with stuck xfers ?

**Comment by belforte** [comment]
@mapellidario why did you do `crab resubmit` ? was it to check what happens ? Or was there a condition which required it ? If jobs failed WN, CRAB should have not issued a xfer request.

**Comment by dciangot** [comment]
So, if this is your task `220524_162113:dmapelli_crab_20220524_182110` I cannot see anythin in kibana around 24th May that is where the first attempt should have happend right? https://cmsweb.cern.ch:8443/scheddmon/0197/dmapelli/220524_162113:dmapelli_crab_20220524_182110/postjob.6.0.txt

Can you may try with a fresh new task? @mapellidario

**Comment by belforte** [comment]
thx Dario, please do not issue any resubmit and if you get jobs with transfer error let's look at them in detail first.
Maybe as simple as "PostJob detected that files were not transferred in 24h and resubmitted" which we know is doomed. Yet it is surprising that Rucio could not get a file to CERN in 24h. We need logs !

**Comment by dciangot** [comment]
Ok. More details from the server(*). Actually no transfers have been triggered, the problem looks like RUCIO does not find a valid path b/w the destination and the source. At this point I ask @ericvaandering consultancy though, is this  related to a missing rucio distance (**)? Maybe some recent changes in RSE sync script?

(*)
2022-06-22 16:21:42,830	root	9	INFO	conveyor-submitter[7/15]: a6b04f309412475a8cfd28065a69599c(user.dmapelli:/store/user/rucio/dmapelli/ruciostagouttest/GenericTTbar/autotest-dmapelli-1655913471/220622_155756/0000/output_5.root): 0 ordered sources: ; 1 dropped due to missing path: T1_DE_KIT_Disk_Temp
2022-06-22 16:21:42,853	root	9	INFO	conveyor-submitter[7/15]: b453b87fb197472190c1541138dfd9ea(user.dmapelli:/store/user/rucio/dmapelli/ruciostagouttest/GenericTTbar/autotest-dmapelli-1655913471/220622_155756/0000/output_2.root): 0 ordered sources: ; 1 dropped due to missing path: T1_RU_JINR_Disk_Temp
2022-06-22 16:21:53,492	root	9	INFO	conveyor-submitter[8/15]: 54ec492237ef46898b699946ec888cd6(user.dmapelli:/store/user/rucio/dmapelli/ruciostagouttest/GenericTTbar/autotest-dmapelli-1655913471/220622_155756/0000/output_3.root): 0 ordered sources: ; 1 dropped due to missing path: T1_RU_JINR_Disk_Temp
2022-06-22 16:21:53,494	root	9	INFO	conveyor-submitter[8/15]: b92c441b19764b2194abd084225bc9ca(user.dmapelli:/store/user/rucio/dmapelli/ruciostagouttest/GenericTTbar/autotest-dmapelli-1655913471/220622_155756/0000/output_8.root): 0 ordered sources: ; 1 dropped due to missing path: T1_DE_KIT_Disk_Temp

(**)
$ rucio-admin rse get-distance T1_DE_KIT_Disk_Temp T2_CH_CERN
No distance set from T1_DE_KIT_Disk_Temp to T2_CH_CERN

**Comment by belforte** [comment]
bravo Diego ! Well.. DT team will need to add those _Temp RSE as well to their monitoring.

**Comment by dciangot** [comment]
It appears that right now only T2 Temp RSEs have been sync (*), so everything running at T1s should fail. I have reported it in ops chat to get a feedback on the best way to proceed.

(*)
https://github.com/dmwm/CMSRucio/blob/master/docker/CMSRucioClient/scripts/sync_sites.sh#L3

**Comment by belforte** [comment]
looking forward, is that chat something crab operators should better be part of ?  I am afraid I do not know which one itis, is it on Slack or MM ?

**Comment by dciangot** [comment]
Is a Rucio slack channel... I definetely think it can be worth for crab people to start getting into it. I'm no slack expert though, I'll need to figure out how to invite you properly :')

**Comment by belforte** [comment]
> Is a Rucio slack channel... I definetely think it can be worth for crab people to start getting into it. I'm no slack expert though, I'll need to figure out how to invite you properly :')

 @ericvaandering  ?

**Comment by dciangot** [comment]
Done :)

**Comment by belforte** [comment]
Looks to me that there is no action for CRAB in here. In the end PostJob can only say: being handled by Rucio. Finding stuck rules (not to mention why they are stuck) is better done from the shell using rucio commands or on Rucio web UI.

If people disagree, please reopen (or better open a new one)
